proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -507,39 +507,1153 @@ locale delete_one = delete_one_conc + delete_one_abs +
+              apply (clarsimp simp: sc_relation_def)
+             apply clarsimp
+            apply (clarsimp simp: objBits_simps)
+           apply simp
+          apply simp
+         apply (wpsimp wp: get_sched_context_exs_valid simp: is_sc_obj_def obj_at_def)
+          apply (rename_tac ko xs; case_tac ko; clarsimp)
+         apply simp
+        apply (wpsimp simp: obj_at_def is_sc_obj_def vs_heap_simps)+
+       apply (wpsimp wp: get_sched_context_no_fail)
+      apply (clarsimp simp: obj_at_def is_sc_obj_def)
+     apply simp
+    apply wpsimp+
+  done
+
+lemma updateReply_reply_sc_update_None_corres:
+  "corres dc (reply_at rp and pspace_aligned and pspace_distinct) \<top>
+    (set_reply_obj_ref reply_sc_update rp None)
+    (updateReply rp (replyNext_update Map.empty))"
+    apply (rule_tac Q="reply_at' rp" in corres_cross_add_guard)
+     apply (fastforce dest!: state_relationD reply_at_cross)
+  apply (clarsimp simp: update_sk_obj_ref_def updateReply_def)
+  apply (rule corres_guard_imp)
+  apply (rule corres_split[OF _ get_reply_corres])
+  apply (rule set_reply_corres)
+  apply (clarsimp simp: reply_relation_def)
+  apply wpsimp+
+  by (clarsimp simp: obj_at'_def replyPrev_same_def)
+
+lemma updateReply_reply_sc_update_Some_corres:
+  "corres dc (reply_at rp and pspace_aligned and pspace_distinct) \<top>
+    (set_reply_obj_ref reply_sc_update rp (Some nr))
+    (updateReply rp (replyNext_update (\<lambda>_. Some (Head nr))))"
+    apply (rule_tac Q="reply_at' rp" in corres_cross_add_guard)
+     apply (fastforce dest!: state_relationD reply_at_cross)
+  apply (clarsimp simp: update_sk_obj_ref_def updateReply_def)
+  apply (rule corres_guard_imp)
+  apply (rule corres_split[OF _ get_reply_corres])
+  apply (rule set_reply_corres)
+  apply (clarsimp simp: reply_relation_def)
+  apply wpsimp+
+  by (clarsimp simp: obj_at'_def replyPrev_same_def)
+
+lemma replyPrevNext_update_commute:
+  "replyPrev_update Map.empty (replyNext_update Map.empty reply)
+   = replyNext_update Map.empty (replyPrev_update Map.empty reply)"
+  by (cases reply; clarsimp)
+
+lemma updateReply_Prev_Next_rewrite:
+  "monadic_rewrite False True (reply_at' rp)
+    (do y <- updateReply rp (replyPrev_update Map.empty);
+         updateReply rp (replyNext_update Map.empty)
+     od)
+    (do y <- updateReply rp (replyNext_update Map.empty);
+         updateReply rp (replyPrev_update Map.empty)
+     od)"
+  apply (clarsimp simp: monadic_rewrite_def)
+  apply (rule monad_state_eqI)
+    apply (find_goal \<open>match conclusion in "_ = _" \<Rightarrow> -\<close>)
+    apply (insert no_fail_updateReply, drule_tac x=rp and y="replyPrev_update Map.empty" in meta_spec2)
+    apply (insert no_fail_updateReply, drule_tac x=rp and y="replyNext_update Map.empty" in meta_spec2)
+    apply (rule; clarsimp simp: in_monad no_fail_def snd_bind split_def)
+     apply (drule (1) use_valid[OF _ updateReply_reply_at'[where P=id, simplified], rotated], fastforce)+
```

Comment:
This line can be replaced with `apply (drule (1) use_valid[OF _ updateReply_reply_at'], fastforce)+`